bundles/org.openhab.binding.surepetcare/src/main/java/org/openhab/binding/surepetcare/internal/SurePetcareAPIHelper.java

```diff
@@ -0,0 +1,474 @@
+     *
+     * @param url the URL
+     * @return a JSON string with the API result
+     * @throws SurePetcareApiException
+     */
+    private String getResultFromApi(String url) throws SurePetcareApiException {
+        Request request = httpClient.newRequest(url).method(HttpMethod.GET);
+        ContentResponse response = executeAPICall(request);
+        String responseData = response.getContentAsString();
+        logger.debug("API execution successful, response: {}", responseData);
+        return responseData;
+    }
+
+    /**
+     * Uses the given request method to send a JSON string to an API.
+     *
+     * @param url the URL
+     * @param method the required request method (POST, PUT etc.)
+     * @param jsonPayload the JSON string
+     * @throws SurePetcareApiException
+     */
+    private void postDataThroughAPI(String url, HttpMethod method, String jsonPayload) throws SurePetcareApiException {
+        logger.debug("postDataThroughAPI URL: {}", url);
+        logger.debug("postDataThroughAPI Payload: {}", jsonPayload);
+        Request request = httpClient.newRequest(url).method(method);
+        request.content(new StringContentProvider(jsonPayload));
+        executeAPICall(request);
+    }
+
+    /**
+     * Uses the given request execute the API call. If it receives an HTTP_UNAUTHORIZED response, it will automatically
+     * login again and retry.
+     *
+     * @param request the Request
+     * @return the response from the API
+     * @throws SurePetcareApiException
+     */
+    private ContentResponse executeAPICall(Request request) throws SurePetcareApiException {
+        int retries = 3;
+        while (retries > 0) {
+            try {
+                setConnectionHeaders(request);
+                ContentResponse response = request.send();
+                if ((response.getStatus() == HttpURLConnection.HTTP_OK)
+                        || (response.getStatus() == HttpURLConnection.HTTP_CREATED)) {
+                    return response;
+                } else {
+                    logger.debug("HTTP Response Code: {}", response.getStatus());
+                    logger.debug("HTTP Response Msg: {}", response.getReason());
+                    if (response.getStatus() == HttpURLConnection.HTTP_UNAUTHORIZED) {
+                        // authentication token has expired, login again and retry
+                        login(username, password);
+                        retries--;
+                    } else {
+                        throw new SurePetcareApiException(
+                                "Http error: " + response.getStatus() + " - " + response.getReason());
+                    }
+                }
+            } catch (AuthenticationException | InterruptedException | ExecutionException | TimeoutException
+                    | ProtocolException e) {
+                throw new SurePetcareApiException("Exception caught during API execution." + e);
```

Comment:
This is a non standard way to use exception chaining, normally you would, pass the parent exception as a cause, however in this case you chose to append the toString representation of the cause to the message.

Reply:
Note however, if you would use the more standard way of chaining exceptions using causes your logging could become problematic because you only seem to print the message and not the stack trace on the first occurrences I spotted.

Reply:
change logging to log exception, not just the exception message.